cilium/cmd/cleanup.go

```diff
@@ -0,0 +1,242 @@
+	if !force {
+		if !confirmCleanup() {
+			return
+		}
+	}
+
+	// ENOENT and similar errors are ignored. Should print all other
+	// errors seen, but continue.  So that one remove function does not
+	// prevent the remaining from running.
+	type cleanupFunc func() error
+	checks := []cleanupFunc{removeAllMaps, unmountFS, removeDirs, removeCNI}
+	for _, clean := range checks {
+		if err := clean(); err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+		}
+	}
+
+	if err := removeRoutesAndLinks(routes, links); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+	}
+}
+
+func showWhatWillBeRemoved(routes map[int]netlink.Route, links map[int]netlink.Link) {
+	fmt.Printf("Warning: Destructive operation. You are about to remove:\n"+
+		"- all BPF maps in %s containing '%s' and '%s'\n"+
+		"- mounted bpffs at %s\n"+
+		"- library code in %s\n"+
+		"- endpoint state in %s\n"+
+		"- CNI configuration at %s\n",
+		bpf.MapPrefixPath(), ciliumLinkPrefix, tunnelMap, bpf.GetMapRoot(),
+		defaults.LibraryPath, defaults.RuntimePath, cniConfig)
+
+	if len(routes) > 0 {
+		fmt.Printf("- routes\n")
+		for _, v := range routes {
+			fmt.Printf("%v\n", v)
+		}
+	}
+
+	if len(links) > 0 {
+		fmt.Println("- links")
+		for _, v := range links {
+			fmt.Printf("%v\n", v)
+		}
+	}
+}
+
+func confirmCleanup() bool {
+	fmt.Printf("The command is non-revertible, do you want to continue [y/N]?\n")
+	var res string
+	fmt.Scanln(&res)
+	return res == "y"
+}
+
+func removeCNI() error {
+	err := os.Remove(cniConfig)
+	if os.IsNotExist(err) {
+		return nil
+	}
+	return err
+}
+
+func removeDirs() error {
+	dirs := []string{defaults.RuntimePath, defaults.LibraryPath}
+	for _, dir := range dirs {
+		// In the unlikely case one of the constants is the root directory, abort.
+		if dir == "/" {
+			return fmt.Errorf("will not remove root directory %s", dir)
+		}
+		if err := os.RemoveAll(dir); err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func unmountFS() error {
+	if !bpf.IsBpffs(bpf.GetMapRoot()) {
+		return nil
+	}
+	return bpf.UnMountFS()
+}
+
+func removeAllMaps() error {
+	mapDir := bpf.MapPrefixPath()
+	maps, err := ioutil.ReadDir(mapDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+
+	for _, m := range maps {
+		name := m.Name()
+		// Skip non Cilium looking maps
+		if !strings.HasPrefix(name, ciliumLinkPrefix) && name != tunnelMap {
+			continue
+		}
+		if err = os.Remove(filepath.Join(mapDir, name)); err != nil {
+			return err
+		}
+		fmt.Printf("removed map %s\n", m.Name())
+	}
+	return err
+}
+
+func linkMatch(linkName string) bool {
+	if strings.HasPrefix(linkName, ciliumLinkPrefix) {
+		return true
+	}
+	return strings.HasPrefix(linkName, hostLinkPrefix) && len(linkName) == hostLinkLen
+}
+
+func findRoutesAndLinks() (map[int]netlink.Route, map[int]netlink.Link, error) {
+	routesToRemove := map[int]netlink.Route{}
+	linksToRemove := map[int]netlink.Link{}
```

Comment:
I don't think we use any of the indexes (the int, here) anywhere. If you preferred to treat these maps as a set you can use `map[*netlink.Route]bool` and `map[*netlink.Link]bool` (or also non-pointer versions, I think). You would then iterate over them in `removeRoutesAndLinks` with only `for route := range routes` etc.

Reply:
Fine as is, thanks.

Reply:
I'm fine with how it's currently done, matter of personal preference.